---
title: "GWAS"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GWAS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Simple association analysis

Using the method of described in gen_sim to simulate genetic data, we can use that data to visualise both association studies and methods of prediction. In this artcle we are going to cover the use and theory of the method GWAS. GWAS, which stands for genome-wide association studies is a method used for, as the name implies, association studies. Meaning that based on large quantities of genetic data we would like to know which places on the genome is critical for developing the the given sickness or trait.  

GWAS is in simple terms a linear regression which for each SNP tries to estimate the casual effect of having a mutation on the probability of having the trait or sickenes. GWAS can therefore be written as the linear regression:

$$
Y \sim \mathcal{N}(X\beta,\Sigma)
$$
Where $Y$ denotes the actual traits of the subjects. $Y$ will be 1 if the subject has the sickness/trait and zero if not. X is the genetic data, if generated by the gen_sim() function, then this equal to G. 


## GWAS using logistic regression

By default the function GWAS() will use a linear regression. The concept of $Y$ as a binary variable open the door for other models. A simple one is the logistic regression which uses the logit function as its activation function to transform the linearregression into a probability curve. This model is also implemented in out GWAs function and can be accessed by setting 'logreg = TRUE'. Though as we wil show the difference in result between logistic and liner regression is minimal.

## Using GWAS

In the following tutorial a file is used which have been generated using the method described in gen_sim. First we load the data. We have here simulated $100,000\times100,000$ datapoints using 'fam=FALSE'. The family structure is unimportant since we are only using the trait of the subject.
```{r eval=FALSE}
library(genstats)
library(bigsnpr)
library(dplyr)
setwd('D:/Data')
genetic_data = snp_attach('example_data.rds')
G=genetic_data$genotypes
y =  genetic_data$fam %>% select('pheno_0') 
beta = genetic_data$map$beta

```

Know we estimate the casual effect of each SNP

```{r eval=FALSE}
gwas_summary = GWAS(G=G, y=y[[1]], p=0.05/100000, logreg = FALSE, ncores=3)
gwas_summary
```

We are using bonferronni correction, ans since we have $100,000$ indiviuals we will use $p=\frac{0.05}{100000}$.

```{r eval=FALSE}
manhattan_plot(gwas_summary, beta, c(0.05/100000, 0.05))
```


```{r eval=FALSE}
scatter_plot(gwas_summary, beta)
```

We see that we catches the most extreme cases. Looking at a powerplot this makes a lot of sense



```{r eval=FALSE}
power_plot(gwas_summary, beta)
```








```{r eval=FALSE, include=FALSE}
tibble('l_estim'=LTFH_fam$l_g_est_0, 'l_g'=LTFH_fam$l_g_0) %>%
  ggplot(aes(x=l_estim, y=l_g)) + geom_point() + geom_abline(slope=1) 


```


```
