---
title: "LT-FH"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GWAS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

A more sofisticated method of analysing the genetic data and finding the causal SNP's is using LT-FH (liability threshold family history). This methods uses a gibbsampler in order to estimate the genetic liability of each subject based on the family history, i.e. if the subject, it's parent or siblings have the given trait or sickness. Read more about gibb sambling in the vignette. Afterwards a linear regression can be used agian as association metahtos



## Using LTFH

In the following tutorial a file is used which have been generated using the method described in gen_sim. First we load the data. We have here simulated $100,000\times100,000$ datapoints using two siblings.
```{r eval=FALSE}
library(genstats)
library(bigsnpr)
library(dplyr)
library(tidyverse)

genetic_data = snp_attach('example_data.rds')
G=genetic_data$genotypes
beta = genetic_data$map$beta

```


```{r eval=FALSE, include=FALSE}
library(genstats)

setwd("~/Uddannelse/Universitet/4. Semester/Dataprojekt/genstats")
genetic_data = snp_attach('smalltest2.rds')

```

We know run LT-FH and as we can see get estimates for each subjects genetic liabilities.
```{r eval=FALSE}
LTFH_est= LTFH(genetic_data, n_sib = 2)
LTFH_est
```
We can know try to estimate the causual effect using the estimated genetic liabilities as our target vector.

```{r eval=FALSE}
LTFH_summary = GWAS(G=G, y=LTFH_est$l_g_0, p=0.05/100000, logreg = FALSE, ncores=3)
LTFH_summary
```

We are using bonferronni correction, ans since we have $100,000$ indiviuals we will use $p=\frac{0.05}{100000}$.

```{r eval=FALSE}
manhatten_plot(LTFH_summary, beta, c(0.05/100000, 0.05))
```


```{r eval=FALSE}
scatter_plot(LTFH_summary, beta)
```





