---
title: "LT-FH"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LT-FH}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


We use the following packages in this vignette.

```{r message=FALSE, warning=FALSE}
library(genstats)
library(bigsnpr)
library(tidyverse)
```


A more sofisticated method of analysing the genetic data and finding the causal SNP's is using LT-FH (liability threshold family history). This methos use a Gibbs sampler in order to estimate the genetic liability of each subject based on the family history, i.e. if the subject, it's parent or siblings have the given trait or sickness. Using this estimate an updated expected value for the genetic liability can be obtained.

LTFH is based on a sampling using Monte Carlo integration. More information about the process can be found here `vignette("Gibbs-sampler")`.
Below is shown three plots with the estimated posterior mean genetic liabilities compared to the standard normal distribution of the genetic liability. The plots shows the outcome of running the gibbs samber for the configurations (0,0,0), (0,0,1) and (1,1,1) respectively.

```{r}
gridExtra::grid.arrange(control_plot(phenos = c(0,0,0), h2 = 0.5, col = "royalblue"),
                        control_plot(phenos = c(0,0,1), h2 = 0.5, col = "mediumorchid"),
                        control_plot(phenos = c(1,1,1), h2 = 0.5, col = "firebrick"), heights = 0.5, name = c("Plots for configuration (0,0,0), (0,0,1), (1,1,1)"))
```

Using the estimates of genetic liabilities instead of the phenotypes as target vector can be very beneficial. We are going to run through an example of this in the next section.


## Using LTFH

In the following tutorial a file is used which have been generated using the method described in 'vignette("Simulation")'. First we load the data. We have here simulated $1e^3\times1e^3$ datapoints using two siblings. In the following 100 SNPs has a casual effect. This data is only used to show case the method and illustrate how the package can be used. If the methods should be used to actual statistical analysis we recomend simulating $1e^5\times1e^5$ datapoints with about 500-1000 casual SNPs. For practical reasons we have not been apple to show case the analysis. We have included pictures to illustrate the results on $1e^5\times1e^5$ data.

```{r}
# The rds file containing simulated data is loaded
genetic_data = snp_attach('genetic_data.rds')

# We save the genotypes and beta for later use
G = genetic_data$genotypes
beta = genetic_data$map$beta

```

We now run LT-FH and as we can see we get the estimates for each subjects genetic liabilities as a new column.

```{r}
LTFH_est = LTFH(data = genetic_data, n_sib = 2)
LTFH_est[1:10,] %>% select(contains(c("0", "pheno"))) %>% 
knitr::kable(., digits = 2, align = "c")
```

How the estimated expected genetic liability match the actual values of l_g_0 can be show by plotting estimated vs true. The plot below is done only with subject and parent to get a more clear plot.
```{r}
genetic_data_nosib = genetic_data

genetic_data_nosib$fam <- genetic_data$fam %>% select(-contains(c("s4", "s3")))

LTFH_est0 = LTFH(data = genetic_data_nosib, n_sib = 0)
  
LTFH_plot(LTFH_est0)
```

This looks great! It seems that the y=x line runs through the mean of each configuration.

The estimated genetic liabilities can now be used as wanted. We are here using them as the target vector in a linear regression in order to estimate the causal effect of each SNP see `vignette("GWAS")`.

```{r}
LTFH_summary = GWAS(G = G, y = LTFH_est$l_g_0, p = 5e-5, logreg = FALSE, ncores = 3)
knitr::kable(LTFH_summary[1:10,], digite = 2, align = "c")
```

In the above table we see the result of the association analysis, where we can se which SNP's is estimated as casual. We are using bonferronni correction to lower the amount of false positive causal SNP. And since we have $1e^3$ indiviuals we will use $p=\frac{0.05}{1000}=5e^{-5}$.


## Outcome for realistic data
We will below give the results of the association analysis on a data set with $1e^5\times1e^5$ datapoints. The process of getting these result is the same as shown above we have here used a p-value with bonferroni correction $p=\frac{0.05}{100000}=5e^{-7}$. The first plot shown below is a manhattan plot. The plot illustrates how well the analysis is to differentiate between casual and non causual snp. Two horisontal lines have been drawn illustrating the tresholds for the p-value. We can se that choosing to use the conservative method of bonferronni correction we describe the data better. Comparing this to the result of GWAS without using estimated liabilities we se that LTFH clearly performs better.


```{r echo=T, results='hide', , fig.show='hide'}
manhattan_plot(gwas_summary = LTFH_summary, beta = beta, thresholds = c(5e-7, 0.05))
```

```{r, echo = FALSE}
knitr::include_graphics("LTFH_manhattan.jpg")
```


This can also be illustrated using scatterplot. Here we see that we are very good at finding estimates af very extreme effects, though it is much more difficult to for the association analysis around the middle. Again this method very much out performs normal `vignette("GWAS")`. 

```{r echo=T, results='hide', , fig.show='hide'}
scatter_plot(gwas_summary = LTFH_summary, beta = beta)
```

```{r, echo = FALSE}
knitr::include_graphics("LTFH_scatter.jpg")
```


At last we include a power plot to show case why we are not able to estimate the smaller causal effects. We simply do not have the power in our test to do so! 

```{r echo=T, results='hide', fig.show='hide'}
power_plot(gwas_summary = LTFH_summary, beta = beta)
```

```{r, echo = FALSE}
knitr::include_graphics("LTFH_power.jpg")
```

### Note on LT-FH on real world data

This performance improvement is, with the simulated data, expected since we know for a fact that the right assumptions about the data is made. If real world data is used it is no more a certainty that the assumptions will hold and we would expect LTFH to perform worse than what is shown here, therefore it could be relevant to update the covariance matrix to suit real world data beta. E. g. by adding some kind of correlation between parent's and child's enviromental liability. In spite of this Hujoel et al.(2020) show that LTFH outperforms the standard GWAS with real world data as well.

