---
title: "Simulation"
vignette: >
  %\VignetteIndexEntry{GWAS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
library(genstats)
library(bigsnpr)
```


## Simulation without family history

The package offers different methods of simulating data. Fist we are going to show how to simply get out the genotypes of N subjects and the genetic and ful liabilities. We are here using a small scale test, but we recomend that if the data should be used for actual statistc analysis that at least $10,000\times10,000$ datapoints are simulated.

```{r eval=FALSE}
gen_sim('simple_sim_example', N=1000, M=1000, C=50, block_size = 100, fam=FALSE)
```

Now there will exist a rds file in the working directoy with the data, which we will load.
```{r eval=FALSE}
simple_data = snp_attach('simple_sim_example.rds')
simple_data
```

Looking at the data we see that the file contains a list. We can here find the genotypes (G), liabilities (fam) and information for each SNP (map).
```{r eval=FALSE}
simple_data$G[(1:5),(1:5)]
simple_data$fam
simple_data$map
```

Using the function dist_check() we can see if the distribution of the full liabilities is as expected, which is standard normal. We are here using af small testset and only expect the the mean and variance to be approximatly 0 and 1.
```{r eval=FALSE}
dist_check(simple_data)
```

We can see that the mean and variance is correct and that the qqplot resembles a normal distribution. 
It is now possible to perfom GWAS for association analysis or use the genotypes for prediction.


## Simulation with family history

Simulation with family history

