---
title: "LT-FH"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LT-FH}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

We use the following packages in this vignette.

```{r}
library(genstats)
library(bigsnpr)
library(tidyverse)
```


A more sofisticated method of analysing the genetic data and finding the causal SNP's is using LT-FH (liability threshold family history). This methods uses a gibbs sampler in order to estimate the genetic liability of each subject based on the family history, i.e. if the subject, it's parent or siblings have the given trait or sickness. Using this estimate an updated expected value for the genetic liability can be obtained.

LTFH is based on a sampling using Monte Carlo integration. More on the on this can be found in the vignette for gibb sampling.
Below is shown three plots with the estimated posterior mean genetic liabilities compared to the standard normal distribution of the genetic liability. The plots shows the outcome of running the gibbs samber for the configurations (0,0,0), (0,0,1) and (1,1,1) respectively.

```{r}
gridExtra::grid.arrange(control_plot(phenos = c(0,0,0), h2 = 0.5, col = "royalblue"),
                        control_plot(phenos = c(0,0,1), h2 = 0.5, col = "mediumorchid"),
                        control_plot(phenos = c(1,1,1), h2 = 0.5, col = "firebrick"), heights = 0.5, name = c("Plots for configuration (0,0,0), (0,0,1), (1,1,1)"))
```

Using the estimates of genetic liabilities can instead of the phenotypes af the target vector can very beneficial. We are going to run through and example of this in the next section.


## Using LTFH

In the following tutorial a file is used which have been generated using the method described in gen_sim. First we load the data. We have here simulated $1e^5\times1e^3$ datapoints using two siblings. In the following alls SNPs has a casual effect. This data is only used to show case the method and illustrate how the package can be used. If the methods should be used to actual statistical analysis we recomend simulating $1e^5\times1e^5$ datapoints with about 500-1000 casual SNPs. For practical reasons we have not been apple to show case the analysis. We have included pictures to illustrate the results on $1e^5\times1e^5$ data.

```{r}
# The rds file containing simulated data is loaded
genetic_data = snp_attach('genetic_data.rds')

# We save the genotypes and beta for later use
G = genetic_data$genotypes
beta = genetic_data$map$beta

```

We know run LT-FH and as we can see get estimates for each subjects genetic liabilities.
```{r}
LTFH_est = LTFH(genetic_data, n_sib = 2)
head(LTFH_est)
```
We can know try to estimate the causual effect using the estimated genetic liabilities as our target vector.

```{r}
LTFH_summary = GWAS(G=G, y=LTFH_est$l_g_0, p=0.05/100000, logreg = FALSE, ncores=3)
head(LTFH_summary)
```

We are using bonferronni correction, ans since we have $100,000$ indiviuals we will use $p=\frac{0.05}{100000}$.

```{r}
manhattan_plot(LTFH_summary, beta, c(0.05/1000, 0.05))
```


```{r}
scatter_plot(LTFH_summary, beta)
```

```{r}
power_plot(LTFH_summary, beta)
```


To illustrate how this would look on real data we have the following pictures:
![Manhattan plot of LTFH used on 100,000 times 100,000 data][LTFH_manhatten.jpg)
![Scatterplot plot of LTFH used on 100,000 times 100,000 data][LTFH_scatterplot.jpg)
![Powerplot of LTFH used on 100,000 times 100,000 data][LTFH_powerplot.jpg)

