<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="genstats">
<title>GWAS • genstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GWAS">
<meta property="og:description" content="genstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">genstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Simulation.html">Simulating data</a>
    <a class="dropdown-item" href="../articles/GWAS.html">GWAS</a>
    <a class="dropdown-item" href="../articles/Gibbs-sampler.html">Gibbs sampler</a>
    <a class="dropdown-item" href="../articles/LT-FH.html">LT-FH</a>
    <a class="dropdown-item" href="../articles/Prediction.html">Prediction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/Holdols/genstats">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>GWAS</h1>
            
      
      
      <div class="d-none name"><code>GWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="simple-association-analysis">Simple association analysis<a class="anchor" aria-label="anchor" href="#simple-association-analysis"></a>
</h2>
<p>Using the method of described in gen_sim to simulate genetic data, we
can use that data to visualise both association studies and methods of
prediction. In this artcle we are going to cover the use and theory of
the method GWAS. GWAS, which stands for genome-wide association studies
is a method used for, as the name implies, association studies. Meaning
that based on large quantities of genetic data we would like to know
which places on the genome is critical for developing the the given
sickness or trait.</p>
<p>GWAS is in simple terms a linear regression which for each SNP tries
to estimate the casual effect of having a mutation on the probability of
having the trait or sickenes. GWAS can therefore be written as the
linear regression:</p>
<p><span class="math display">\[
Y \sim \mathcal{N}(X\beta,\Sigma)
\]</span> Where <span class="math inline">\(Y\)</span> denotes the
actual traits of the subjects. <span class="math inline">\(Y\)</span>
will be 1 if the subject has the sickness/trait and zero if not. X is
the genetic data, if generated by the gen_sim() function, then this
equal to G.</p>
</div>
<div class="section level2">
<h2 id="gwas-using-logistic-regression">GWAS using logistic regression<a class="anchor" aria-label="anchor" href="#gwas-using-logistic-regression"></a>
</h2>
<p>By default the function GWAS() will use a linear regression. The
concept of <span class="math inline">\(Y\)</span> as a binary variable
open the door for other models. A simple one is the logistic regression
which uses the logit function as its activation function to transform
the linearregression into a probability curve. This model is also
implemented in our GWAs function and can be accessed by setting ‘logreg
= TRUE’.</p>
</div>
<div class="section level2">
<h2 id="using-gwas">Using GWAS<a class="anchor" aria-label="anchor" href="#using-gwas"></a>
</h2>
<p>In the following tutorial a file is used which have been generated
using the method described in gen_sim. First we load the data. We have
here simulated <span class="math inline">\(100,000\times100,000\)</span>
datapoints using ‘fam=FALSE’. The family structure is unimportant since
we are only using the trait of the subject.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://holdols.github.io/genstats" class="external-link">genstats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genetic_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">'genetic_data.rds'</span><span class="op">)</span>
<span class="va">G</span><span class="op">=</span><span class="va">genetic_data</span><span class="op">$</span><span class="va">genotypes</span>
<span class="va">y</span> <span class="op">=</span>  <span class="va">genetic_data</span><span class="op">$</span><span class="va">fam</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="st">'pheno_0'</span><span class="op">)</span> 
<span class="va">beta</span> <span class="op">=</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">beta</span></code></pre></div>
<p>Now we estimate the casual effect of each SNP</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gwas_summary</span> <span class="op">=</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>G<span class="op">=</span><span class="va">G</span>, y<span class="op">=</span><span class="va">y</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, p<span class="op">=</span><span class="fl">0.05</span><span class="op">/</span><span class="fl">100000</span>, logreg <span class="op">=</span> <span class="cn">FALSE</span>, ncores<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_summary</span><span class="op">)</span>
<span class="co">#&gt;          estim     std.err      score     p_vals causal_estimate</span>
<span class="co">#&gt; 1 -0.005987535 0.004443138 -1.3475914 0.17782039               0</span>
<span class="co">#&gt; 2 -0.006664683 0.003534050 -1.8858483 0.05934436               0</span>
<span class="co">#&gt; 3 -0.009804117 0.007506549 -1.3060750 0.19155709               0</span>
<span class="co">#&gt; 4 -0.002721957 0.004071671 -0.6685111 0.50382286               0</span>
<span class="co">#&gt; 5 -0.003998108 0.003511870 -1.1384556 0.25495758               0</span>
<span class="co">#&gt; 6  0.005265846 0.008903985  0.5914033 0.55426363               0</span></code></pre></div>
<p>We are using bonferronni correction, and since we have <span class="math inline">\(100,000\)</span> indiviuals we will use <span class="math inline">\(p=\frac{0.05}{100000}\)</span>.</p>
</div>
<div class="section level2">
<h2 id="perfomance">Perfomance<a class="anchor" aria-label="anchor" href="#perfomance"></a>
</h2>
<p>To illustrate the perfomance of GWAS The first plot shown below is a
manhattan plot. The plot illustrates how well the analysis is to
differentiate between casual and non causual snp. Two horisontal lines
have been drawn illustrating the tresholds for the p-value.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattan_plot.html">manhattan_plot</a></span><span class="op">(</span><span class="va">gwas_summary</span>, <span class="va">beta</span>, treshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5e-7</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="gwas_manhattan.jpg"><!-- --></p>
<p>This can also be illustrated using scatterplot. Here we see that we
are good at finding estimates with very extreme effects, though it is
much more difficult to for the association analysis around the
middle.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/scatter_plot.html">scatter_plot</a></span><span class="op">(</span><span class="va">gwas_summary</span>, <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="gwas_scatter.jpg"><!-- --></p>
<p>We see that we catches the most extreme cases. Looking at a powerplot
we see that we do not have the power to catch the less extreme
cases.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/power_plot.html">power_plot</a></span><span class="op">(</span><span class="va">gwas_summary</span>, <span class="va">beta</span><span class="op">)</span> </code></pre></div>
<p><img src="gwas_power.jpg"><!-- --></p>
</div>
<div class="section level2">
<h2 id="logistic-regression">Logistic regression<a class="anchor" aria-label="anchor" href="#logistic-regression"></a>
</h2>
<p>Since our target variable is binary it would seem reasonable to run
logistic regression instead of linear.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gwaslog_summary</span> <span class="op">=</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>G<span class="op">=</span><span class="va">G</span>, y<span class="op">=</span><span class="va">y</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, p<span class="op">=</span><span class="fl">0.05</span><span class="op">/</span><span class="fl">100000</span>, logreg <span class="op">=</span> <span class="cn">TRUE</span>, ncores<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gwas_summary</span><span class="op">)</span>
<span class="co">#&gt;          estim     std.err      score     p_vals causal_estimate</span>
<span class="co">#&gt; 1 -0.005987535 0.004443138 -1.3475914 0.17782039               0</span>
<span class="co">#&gt; 2 -0.006664683 0.003534050 -1.8858483 0.05934436               0</span>
<span class="co">#&gt; 3 -0.009804117 0.007506549 -1.3060750 0.19155709               0</span>
<span class="co">#&gt; 4 -0.002721957 0.004071671 -0.6685111 0.50382286               0</span>
<span class="co">#&gt; 5 -0.003998108 0.003511870 -1.1384556 0.25495758               0</span>
<span class="co">#&gt; 6  0.005265846 0.008903985  0.5914033 0.55426363               0</span></code></pre></div>
<p>To check the perfomance we use the same method as above.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattan_plot.html">manhattan_plot</a></span><span class="op">(</span><span class="va">gwaslog_summary</span>, <span class="va">beta</span>, thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5e-7</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="GWASlog_manhattan.jpg"><!-- --></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/scatter_plot.html">scatter_plot</a></span><span class="op">(</span><span class="va">gwaslog_summary</span>, <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="GWASlog_scatter.jpg"><!-- --></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/power_plot.html">power_plot</a></span><span class="op">(</span><span class="va">gwas_summary</span>, <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="GWASlog_power.jpg"><!-- --> According to the scatter plot
it would seem logistic regression is better at estimating the effect
size of each casual SNP, from the power plot however the logistic
regression seem weaker.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Louise Mølgaard Laustsen, Ole Sahlholdt Hansen, Holdols.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
