<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="genstats">
<title>LT-FH • genstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LT-FH">
<meta property="og:description" content="genstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">genstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Simulation.html">Simulating data</a>
    <a class="dropdown-item" href="../articles/GWAS.html">GWAS</a>
    <a class="dropdown-item" href="../articles/Gibbs-sampler.html">Gibbs sampler</a>
    <a class="dropdown-item" href="../articles/LT-FH.html">LT-FH</a>
    <a class="dropdown-item" href="../articles/Prediction.html">Prediction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/Holdols/genstats">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>LT-FH</h1>
            
      
      
      <div class="d-none name"><code>LT-FH.Rmd</code></div>
    </div>

    
    
<p>We use the following packages in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://holdols.github.io/genstats" class="external-link">genstats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
<p>A more sofisticated method of analysing the genetic data and finding
the causal SNP’s is using LT-FH (liability threshold family history).
This methods uses a gibbs sampler in order to estimate the genetic
liability of each subject based on the family history, i.e. if the
subject, it’s parent or siblings have the given trait or sickness. Using
this estimate an updated expected value for the genetic liability can be
obtained.</p>
<p>LTFH is based on a sampling using Monte Carlo integration. More
information about the process can be found her
‘vignette(“Gibbs-smapler”)’. Below is shown three plots with the
estimated posterior mean genetic liabilities compared to the standard
normal distribution of the genetic liability. The plots shows the
outcome of running the gibbs samber for the configurations (0,0,0),
(0,0,1) and (1,1,1) respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"royalblue"</span><span class="op">)</span>,
                        <span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"mediumorchid"</span><span class="op">)</span>,
                        <span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span>, heights <span class="op">=</span> <span class="fl">0.5</span>, name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plots for configuration (0,0,0), (0,0,1), (1,1,1)"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="LT-FH_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Using the estimates of genetic liabilities can instead of the
phenotypes af the target vector can very beneficial. We are going to run
through and example of this in the next section.</p>
<div class="section level2">
<h2 id="using-ltfh">Using LTFH<a class="anchor" aria-label="anchor" href="#using-ltfh"></a>
</h2>
<p>In the following tutorial a file is used which have been generated
using the method described in ‘vignette(“Simulation”)’. First we load
the data. We have here simulated <span class="math inline">\(1e^3\times1e^3\)</span> datapoints using two
siblings. In the following 100 SNPs has a casual effect. This data is
only used to show case the method and illustrate how the package can be
used. If the methods should be used to actual statistical analysis we
recomend simulating <span class="math inline">\(1e^5\times1e^5\)</span>
datapoints with about 500-1000 casual SNPs. For practical reasons we
have not been apple to show case the analysis. We have included pictures
to illustrate the results on <span class="math inline">\(1e^5\times1e^5\)</span> data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The rds file containing simulated data is loaded</span>
<span class="va">genetic_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">'genetic_data.rds'</span><span class="op">)</span>

<span class="co"># We save the genotypes and beta for later use</span>
<span class="va">G</span> <span class="op">=</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">genotypes</span>
<span class="va">beta</span> <span class="op">=</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">beta</span></code></pre></div>
<p>We know run LT-FH and as we can see we get the estimates for each
subjects genetic liabilities as a new column.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LTFH_est</span> <span class="op">=</span> <span class="fu"><a href="../reference/LTFH.html">LTFH</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">genetic_data</span>, n_sib <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LTFH_est</span><span class="op">)</span>
<span class="co">#&gt;   FID      l_f_0      l_g_0 pheno_0      l_f_p1     l_g_p1 pheno_p1      l_f_p2</span>
<span class="co">#&gt; 1   1 -0.3384807 -0.7064320       0  0.86542968 -0.4871935        0  0.10331741</span>
<span class="co">#&gt; 2   2  1.4324619  0.4944826       0  0.77872081  1.1045429        0  0.60545412</span>
<span class="co">#&gt; 3   3 -1.0801217 -0.3403101       0 -0.09001458  1.0320972        0 -0.22832639</span>
<span class="co">#&gt; 4   4  1.2669931  0.5825709       0 -1.24249284 -0.3447710        0  0.86003095</span>
<span class="co">#&gt; 5   5  0.3125711  1.0115020       0  1.74724276  1.3767706        1  0.87617390</span>
<span class="co">#&gt; 6   6 -0.5124047  0.1266217       0 -0.42144219 -0.1451376        0 -0.02902428</span>
<span class="co">#&gt;       l_g_p2 pheno_p2     l_f_s3     l_g_s3 pheno_s3      l_f_s4      l_g_s4</span>
<span class="co">#&gt; 1  0.2801320        0  0.5112359  0.2131810        0 -1.50060122 -0.11247134</span>
<span class="co">#&gt; 2  0.4630299        0  2.0563382  1.3112339        1  1.35933840  1.00005536</span>
<span class="co">#&gt; 3 -0.7587923        0 -0.1088843 -0.1558196        0 -0.72531820 -0.79113094</span>
<span class="co">#&gt; 4  1.2059687        0  1.5115114  0.4552358        0  0.45513811  0.08464925</span>
<span class="co">#&gt; 5  0.3278623        0  0.4358816  0.9360237        0  0.70031543  1.14029948</span>
<span class="co">#&gt; 6 -0.6010279        0 -0.4068985 -0.5806311        0  0.09545961  0.90836585</span>
<span class="co">#&gt;   pheno_s4  l_g_est_0</span>
<span class="co">#&gt; 1        0 -0.1257546</span>
<span class="co">#&gt; 2        0  0.3033850</span>
<span class="co">#&gt; 3        0 -0.1257546</span>
<span class="co">#&gt; 4        0 -0.1257546</span>
<span class="co">#&gt; 5        0  0.3431188</span>
<span class="co">#&gt; 6        0 -0.1257546</span></code></pre></div>
<p>The estimated genetic liabilities can now be used as wanted. We are
here using them as the target vector in a linear regression in order to
estimate the causal effect of each SNP.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LTFH_summary</span> <span class="op">=</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, y <span class="op">=</span> <span class="va">LTFH_est</span><span class="op">$</span><span class="va">l_g_0</span>, p <span class="op">=</span> <span class="fl">5e-5</span>, logreg <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LTFH_summary</span><span class="op">)</span>
<span class="co">#&gt;          estim    std.err      score      p_vals causal_estimate</span>
<span class="co">#&gt; 1  0.002007187 0.01383549  0.1450753 0.884654371               0</span>
<span class="co">#&gt; 2 -0.010628664 0.01100513 -0.9657913 0.334171958               0</span>
<span class="co">#&gt; 3  0.003329657 0.02337451  0.1424482 0.886728893               0</span>
<span class="co">#&gt; 4 -0.041767248 0.01267104 -3.2962767 0.000983169               0</span>
<span class="co">#&gt; 5  0.010018953 0.01093488  0.9162381 0.359564110               0</span>
<span class="co">#&gt; 6 -0.026468634 0.02772285 -0.9547587 0.339722810               0</span></code></pre></div>
<p>In the above table we se the result of the association analysis,
where we can se which SNP’s is estimated as casual. We are here using
bonferronni correction to lower the amont of false positive causal SNP.
And since we have <span class="math inline">\(1e^3\)</span> indiviuals
we will use <span class="math inline">\(p=\frac{0.05}{1000}=5e^{-5}\)</span>. We can se
the result of our</p>
</div>
<div class="section level2">
<h2 id="outcome-for-realistic-data">Outcome for realistic data<a class="anchor" aria-label="anchor" href="#outcome-for-realistic-data"></a>
</h2>
<p>We will below give the results of the association analysis on a data
set with <span class="math inline">\(1e^5\times1e^5\)</span> datapoint.
The process of getting these result is the same as shown above we have
here used a p-value with bonferroni correction <span class="math inline">\(p=\frac{0.05}{100000}=5e^{-7}\)</span>. The first
plot shown below is a manhattan plot. The plot illustrates how well the
analysis is to differentiate between casual and non causual snp. Two
horisontal lines have been drawn illustrating the tresholds for the
p-value. We can se that choosing to use the conservative method of
bonferronni correction we describe the data better. Comparing this to
the result of GWAS without using estimated liabilities we se that LTFH
clearly performs better.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattan_plot.html">manhattan_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span>, treshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5e-7</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_manhattan.jpg"><!-- --></p>
<p>This can also be illustrated using scatterplot. Here we see that we
are very good at finding estimates af very extreme effects, though it is
much more difficult to for the association analysis around the middle.
Agian this method very much out performs normal ‘vignette(“GWAS”)’.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/scatter_plot.html">scatter_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_scatter.jpg"><!-- --></p>
<p>At last we include a power plot to show case why we are not able to
estimate the smaller causal effects. We simply do not have the power in
our test to do so!</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/power_plot.html">power_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_power.jpg"><!-- --></p>
<div class="section level3">
<h3 id="note-on-lt-fh-on-real-world-data">Note on LT-FH on real world data<a class="anchor" aria-label="anchor" href="#note-on-lt-fh-on-real-world-data"></a>
</h3>
<p>This performance improvement is, with the simulated data, expected
since we know for a fact that the right assumptions about the data is
made. If real world data is used it is no more a certainty that the
assumptions will hold and we would expect LTFH to perform worse than
what is shown here, therefore it could be relevant to update the
covariance matrix to suit real world data beta. E. g. by adding some
kind of correlation between parent’s and child’s enviromental liability.
In spite of this Hujoel et al.(2020) show that LTFH outperforms the
standard GWAS with real world data as well.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Louise Mølgaard Laustsen, Ole Sahlholdt Hansen, Holdols.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
