<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="genstats">
<title>LT-FH • genstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LT-FH">
<meta property="og:description" content="genstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">genstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Simulation.html">Simulating data</a>
    <a class="dropdown-item" href="../articles/GWAS.html">GWAS</a>
    <a class="dropdown-item" href="../articles/Gibbs-sampler.html">Gibbs sampler</a>
    <a class="dropdown-item" href="../articles/LT-FH.html">LT-FH</a>
    <a class="dropdown-item" href="../articles/Prediction.html">Prediction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/Holdols/genstats">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>LT-FH</h1>
            
      
      
      <div class="d-none name"><code>LT-FH.Rmd</code></div>
    </div>

    
    
<p>We use the following packages in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://holdols.github.io/genstats" class="external-link">genstats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
<p>A more sophisticated method of analyzing the genetic data and finding
the causal SNPs is using LT-FH (liability threshold family history).
This method use a Gibbs sampler in order to estimate the genetic
liability of each subject based on the family history, i.e. if the
subject, it’s parent or siblings have the given trait or sickness. Using
this estimate an updated expected value for the genetic liability can be
obtained.</p>
<p>LTFH is based on a sampling using Monte Carlo integration. More
information about the process can be found here
<code><a href="../articles/Gibbs-sampler.html">vignette("Gibbs-sampler")</a></code>. Below is shown three plots with
the estimated posterior mean genetic liabilities compared to the
standard normal distribution of the genetic liability. The plots shows
the outcome of running the Gibbs sampler for the configurations (0,0,0),
(0,0,1) and (1,1,1) respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"royalblue"</span><span class="op">)</span>,
                        <span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"mediumorchid"</span><span class="op">)</span>,
                        <span class="fu"><a href="../reference/control_plot.html">control_plot</a></span><span class="op">(</span>phenos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, h2 <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span>,
                        top <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plots for configuration (0,0,0), (0,0,1), (1,1,1)"</span><span class="op">)</span>,
                        ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="LT-FH_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Using the estimates of genetic liabilities instead of the phenotypes
as target vector can be very beneficial. We are going to run through an
example of this in the next section.</p>
<div class="section level2">
<h2 id="using-ltfh">Using LTFH<a class="anchor" aria-label="anchor" href="#using-ltfh"></a>
</h2>
<p>In the following tutorial a file is used which have been generated
using the method described in <code><a href="../articles/Simulation.html">vignette("Simulation")</a></code>. First
we load the data. We have here simulated <span class="math inline">\(1e^4\times1e^4\)</span> data points using two
siblings. In the following 100 SNPs has a casual effect. This data is
only used to show case the method and illustrate how the package can be
used. If the methods should be used to actual statistical analysis we
recommend simulating <span class="math inline">\(1e^5\times1e^5\)</span>
data points with about 500-1000 casual SNPs. For practical reasons we
have not been apple to show case the analysis. We have included pictures
to illustrate the results on <span class="math inline">\(1e^5\times1e^5\)</span> data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The rds file containing simulated data is loaded</span>
<span class="va">genetic_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">'genetic_data.rds'</span><span class="op">)</span>

<span class="co"># We save the genotypes and beta for later use</span>
<span class="va">G</span> <span class="op">=</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">genotypes</span>
<span class="va">beta</span> <span class="op">=</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">beta</span></code></pre></div>
<p>We now run LT-FH and as we can see we get the estimates for each
subjects genetic liabilities as a new column.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LTFH_est</span> <span class="op">=</span> <span class="fu"><a href="../reference/LTFH.html">LTFH</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">genetic_data</span>, n_sib <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">LTFH_est</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span>, <span class="st">"pheno"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">.</span>, digits <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"c"</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="9%">
<col width="9%">
<col width="12%">
<col width="14%">
<col width="13%">
<col width="13%">
<col width="13%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center">l_f_0</th>
<th align="center">l_g_0</th>
<th align="center">pheno_0</th>
<th align="center">l_g_est_0</th>
<th align="center">pheno_p1</th>
<th align="center">pheno_p2</th>
<th align="center">pheno_s3</th>
<th align="center">pheno_s4</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">-0.34</td>
<td align="center">-0.71</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">1.43</td>
<td align="center">0.49</td>
<td align="center">0</td>
<td align="center">0.32</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">-1.08</td>
<td align="center">-0.34</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">1.27</td>
<td align="center">0.58</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">0.31</td>
<td align="center">1.01</td>
<td align="center">0</td>
<td align="center">0.34</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.51</td>
<td align="center">0.13</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">0.32</td>
<td align="center">0.10</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">0.58</td>
<td align="center">0.77</td>
<td align="center">0</td>
<td align="center">0.85</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center">0.23</td>
<td align="center">-0.26</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-1.47</td>
<td align="center">-0.98</td>
<td align="center">0</td>
<td align="center">-0.14</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>How the estimated expected genetic liability match the actual values
of l_g_0 can be show by plotting estimated vs true. The plot below is
done only with subject and parent to get a more clear plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genetic_data_nosib</span> <span class="op">=</span> <span class="va">genetic_data</span>

<span class="va">genetic_data_nosib</span><span class="op">$</span><span class="va">fam</span> <span class="op">&lt;-</span> <span class="va">genetic_data</span><span class="op">$</span><span class="va">fam</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s4"</span>, <span class="st">"s3"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">LTFH_est0</span> <span class="op">=</span> <span class="fu"><a href="../reference/LTFH.html">LTFH</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">genetic_data_nosib</span>, n_sib <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  
<span class="fu"><a href="../reference/LTFH_plot.html">LTFH_plot</a></span><span class="op">(</span><span class="va">LTFH_est0</span><span class="op">)</span></code></pre></div>
<p><img src="LT-FH_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>This looks great! It seems that the y=x line runs through the mean of
each configuration.</p>
<p>The estimated genetic liabilities can now be used as wanted. We are
here using them as the target vector in a linear regression in order to
estimate the causal effect of each SNP see
<code><a href="../articles/GWAS.html">vignette("GWAS")</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">LTFH_summary</span> <span class="op">=</span> <span class="fu"><a href="../reference/GWAS.html">GWAS</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>, y <span class="op">=</span> <span class="va">LTFH_est</span><span class="op">$</span><span class="va">l_g_0</span>, p <span class="op">=</span> <span class="fl">5e-6</span>, logreg <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">LTFH_summary</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span>, digite <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"c"</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="center">estim</th>
<th align="center">std.err</th>
<th align="center">score</th>
<th align="center">p_vals</th>
<th align="center">causal_estimate</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">0.0020072</td>
<td align="center">0.0138355</td>
<td align="center">0.1450753</td>
<td align="center">0.8846544</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.0106287</td>
<td align="center">0.0110051</td>
<td align="center">-0.9657913</td>
<td align="center">0.3341720</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">0.0033297</td>
<td align="center">0.0233745</td>
<td align="center">0.1424482</td>
<td align="center">0.8867289</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.0417672</td>
<td align="center">0.0126710</td>
<td align="center">-3.2962767</td>
<td align="center">0.0009832</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">0.0100190</td>
<td align="center">0.0109349</td>
<td align="center">0.9162381</td>
<td align="center">0.3595641</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.0264686</td>
<td align="center">0.0277229</td>
<td align="center">-0.9547587</td>
<td align="center">0.3397228</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">-0.0031987</td>
<td align="center">0.0110099</td>
<td align="center">-0.2905253</td>
<td align="center">0.7714204</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.0019586</td>
<td align="center">0.0137008</td>
<td align="center">-0.1429572</td>
<td align="center">0.8863268</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">0.0143384</td>
<td align="center">0.0109367</td>
<td align="center">1.3110289</td>
<td align="center">0.1898781</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">-0.0118281</td>
<td align="center">0.0128857</td>
<td align="center">-0.9179204</td>
<td align="center">0.3586827</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>In the above table we see the result of the association analysis,
where we can see which SNPs is estimated as casual. We are using
Bonferroni correction to lower the amount of false positive causal SNP.
And since we have <span class="math inline">\(10^4\)</span> individuals
we will use <span class="math inline">\(p==5e\cdot 10^{-6}\)</span>.</p>
</div>
<div class="section level2">
<h2 id="outcome-for-realistic-data">Outcome for realistic data<a class="anchor" aria-label="anchor" href="#outcome-for-realistic-data"></a>
</h2>
<p>We will below give the results of the association analysis on a data
set with <span class="math inline">\(10^5\times10^5\)</span> data
points. The process of getting these result is the same as shown above
we have here used a p-value with Bonferroni correction <span class="math inline">\(p=5\cdot 10^{-7}\)</span>. The first plot shown
below is a Manhattan plot. The plot illustrates how well the analysis is
to differentiate between casual and non causal SNP. Two horisontal lines
have been drawn illustrating the thresholds for the p-value. We can see
that choosing to use the conservative method of Bonferroni correction we
describe the data better. Comparing this to the result of GWAS without
using estimated liabilities we see that LTFH clearly performs
better.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattan_plot.html">manhattan_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span>, thresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5e-7</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_manhattan.jpg"><!-- --></p>
<p>This can also be illustrated using scatter plot. Here we see that we
are very good at finding estimates of very extreme effects, though it is
much more difficult to for the association analysis around the middle.
Again this method very much outperforms normal
<code><a href="../articles/GWAS.html">vignette("GWAS")</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/scatter_plot.html">scatter_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_scatter.jpg"><!-- --></p>
<p>At last we include a power plot to show case why we are not able to
estimate the smaller causal effects. We simply do not have the power in
our test to do so!</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/power_plot.html">power_plot</a></span><span class="op">(</span>gwas_summary <span class="op">=</span> <span class="va">LTFH_summary</span>, beta <span class="op">=</span> <span class="va">beta</span><span class="op">)</span></code></pre></div>
<p><img src="LTFH_power.jpg"><!-- --></p>
<div class="section level3">
<h3 id="note-on-lt-fh-on-real-world-data">Note on LT-FH on real world data<a class="anchor" aria-label="anchor" href="#note-on-lt-fh-on-real-world-data"></a>
</h3>
<p>This performance improvement is, with the simulated data, expected
since we know for a fact that the right assumptions about the data is
made. If real world data is used it is no more a certainty that the
assumptions will hold and we would expect LTFH to perform worse than
what is shown here, therefore it could be relevant to update the
covariance matrix to suit real world data beta. E. g. by adding some
kind of correlation between parent’s and child’s environmental
liability. In spite of this Hujoel et al.(2020) show that LTFH
outperforms the standard GWAS with real world data as well.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Louise Mølgaard Laustsen, Ole Sahlholdt Hansen, Holdols.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
