<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="genstats">
<title>Prediction • genstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prediction">
<meta property="og:description" content="genstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">genstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Simulation.html">Simulating data</a>
    <a class="dropdown-item" href="../articles/GWAS.html">GWAS</a>
    <a class="dropdown-item" href="../articles/Gibbs-sampler.html">Gibbs sampler</a>
    <a class="dropdown-item" href="../articles/LT-FH.html">LT-FH</a>
    <a class="dropdown-item" href="../articles/Prediction.html">Prediction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/Holdols/genstats">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Prediction</h1>
            
      
      
      <div class="d-none name"><code>Prediction.Rmd</code></div>
    </div>

    
    
<p>We are in this vignette showing the possibilities of using the result
of association analysis in order to predict occurence of the trait by
looking at genetic data. We are in this analysis only looking at genetic
data, which normally would not be case. In predicting the prevelance of
trait/sickness many other factors will have to be taken into account.
This could be sex, age, BMI or other potential confounding variables. We
are though going to show that it is possible to create a stronger model
through the use of genetic data if there of course is a strong enough
heritability. We are therefor going to use the method on <span class="math inline">\(1e^5\times1e^3\)</span> data with 500 casual SNPs.
This is not realistic data, but i will show the potential of the
method.</p>
<p>We use the following packages in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://holdols.github.io/genstats" class="external-link">genstats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="introduction-to-prs">Introduction to PRS<a class="anchor" aria-label="anchor" href="#introduction-to-prs"></a>
</h2>
<p>In prediction we use a trainings set and a test set. We are in the
following section going to show a method which uses LT-FH estimates as
its target vectors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Loading our train and test set</span>
<span class="va">train_set</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">"genetic_data.rds"</span><span class="op">)</span>
<span class="va">test_set</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">"genetic_data_test.rds"</span><span class="op">)</span>

<span class="co"># We save the genotypes and target vector</span>
<span class="va">G</span> <span class="op">=</span> <span class="va">train_set</span><span class="op">$</span><span class="va">genotypes</span>
<span class="va">est</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LTFH.html">LTFH</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">train_set</span>, n_sib <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"est"</span><span class="op">)</span><span class="op">)</span>
<span class="va">y_ltfh</span> <span class="op">=</span> <span class="va">est</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>Using PRS we want to find which threshold for the p-value describes
the data best. We are therefore using cross validation. The following
function outputs PRS estimates for the validation set for each fold in
the training data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">prs</span> <span class="op">=</span> <span class="fu"><a href="../reference/PRS_cross.html">PRS_cross</a></span><span class="op">(</span>train_data <span class="op">=</span> <span class="va">train_set</span>, y <span class="op">=</span> <span class="va">y_ltfh</span>, cross_folds <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre></div>
<p>We use these scores to evaluate the threshold.</p>
<p>The following plot has been evaluated on <span class="math inline">\(1e^5\times1e^3\)</span> data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/prs_plot.html">prs_plot</a></span><span class="op">(</span>PRS <span class="op">=</span> <span class="va">prs</span>, <span class="va">train_set</span>, <span class="st">'AUC'</span><span class="op">)</span></code></pre></div>
<p><img src="m1_auc.jpg"><!-- --><img src="m2_auc.jpg"><!-- --></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/prs_plot.html">prs_plot</a></span><span class="op">(</span>PRS <span class="op">=</span> <span class="va">prs</span>, <span class="va">train_set</span>, <span class="st">'MSE'</span><span class="op">)</span></code></pre></div>
<p><img src="m1_mse.jpg"><!-- --><img src="m2_mse.jpg"><!-- --></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/prs_plot.html">prs_plot</a></span><span class="op">(</span>PRS <span class="op">=</span> <span class="va">prs</span>, <span class="va">train_set</span>, <span class="st">'R2'</span><span class="op">)</span></code></pre></div>
<p><img src="m1_r2.jpg"><!-- --><img src="m2_r2.jpg"><!-- --></p>
<p>Looking at the above plot we find that the optimal threshold for the
p-value is around 3. Next we would like to know for which value of PRS
as a desicion boundary we could expect the best result. We not
considering how may false negatives we experience. It is much more
important that the amount of false positives is very low. Like GWAS or
LT-FH would we prefer a much more conservative prediction. We are using
cross validation again with the following output for the <span class="math inline">\(^5\times1e^3\)</span> matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/decision_cross.html">decision_cross</a></span><span class="op">(</span>train_data <span class="op">=</span> <span class="va">train_set</span>, y <span class="op">=</span> <span class="va">y_ltfh</span>, cross_folds <span class="op">=</span> <span class="fl">4</span>, bounds<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.01</span><span class="op">)</span>, thr <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="decision.jpg"><!-- --></p>
<p>We will choose a boundary around 0.25. We are now ready to create a
prediction.</p>
</div>
<div class="section level2">
<h2 id="prediction-using-prs">Prediction using PRS<a class="anchor" aria-label="anchor" href="#prediction-using-prs"></a>
</h2>
<p>Using this we can create our model. We are only using a linear
regression, since we are using LT-FH estimates as our target vector.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ltfh_model</span> <span class="op">=</span> <span class="fu"><a href="../reference/pred_model.html">pred_model</a></span><span class="op">(</span>train_data <span class="op">=</span> <span class="va">train_set</span>, y <span class="op">=</span> <span class="va">y_ltfh</span>, thr <span class="op">=</span> <span class="fl">3</span>, LogReg <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="prs_est_train.jpg"><!-- --></p>
<p>And creating our predictions</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pred</span> <span class="op">=</span> <span class="fu"><a href="../reference/prediction.html">prediction</a></span><span class="op">(</span>test_data <span class="op">=</span> <span class="va">test_set</span>, gwas <span class="op">=</span> <span class="va">ltfh_model</span>, thr <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>We know use the value we found for the decision boundery and we have
the following result.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">preds</span> <span class="op">=</span> <span class="op">(</span><span class="va">pred</span> <span class="op">&gt;</span> <span class="fl">0.25</span><span class="op">)</span><span class="op">-</span><span class="fl">0</span>
<span class="va">confusion_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="st">'Predicted'</span> <span class="op">=</span> <span class="va">preds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="st">'Actual'</span><span class="op">=</span> <span class="va">test_set</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">pheno_0</span><span class="op">)</span><span class="op">)</span>

<span class="va">confusion_matrix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Predicted</span>, y <span class="op">=</span> <span class="va">Actual</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">n</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%1.0f"</span>, <span class="va">n</span><span class="op">)</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>high <span class="op">=</span> <span class="st">"firebrick"</span>, low <span class="op">=</span> <span class="st">'dodgerblue3'</span>, trans<span class="op">=</span><span class="st">'pseudo_log'</span><span class="op">)</span></code></pre></div>
<p><img src="result_decision.jpg"><!-- --></p>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>We will in this section explain the process behind the above result.
We can only expect to be able to predict the genetic liability. If a
subject has a very low genetic liability, but high envirromental
liability we can of course not catch this with PRS. We are therefore
going to plot the PRS agianst the genetic liablity.</p>
<p><img src="prs_true_train.jpg"><!-- --></p>
<p>We can see from the above plot that these two values are very
correlated. It is therefore reasonable to add PRS as a controlling
variable when prediction studies is performed. We can also plot the PRS
of the test set.</p>
<p><img src="prs_true_test.jpg"><!-- --></p>
<p>As excpected boundary between cases and non cases is a bit less
clear, but we still se a very high correlation.</p>
</div>
<div class="section level2">
<h2 id="using-phenotypes-instead-of-lt-fh-estimates">USing phenotypes instead of LT-FH estimates<a class="anchor" aria-label="anchor" href="#using-phenotypes-instead-of-lt-fh-estimates"></a>
</h2>
<p>We have in the above used LT-FH estimaes in our prediction. We have
done this as we believe this is a superior method with higher power and
more accuracy. One argument for this is that we trough PRS cant expect
to predict the cases, but we can predict who has a high genetic
liability.</p>
<p>Using the same method as above we can see that the correlation
between the genetic liability and PRS found with linear regression or
logistic regression using phenotypes is much lower.</p>
<p><img src="prs_pheno.jpg"><!-- --></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Louise Mølgaard Laustsen, Ole Sahlholdt Hansen, Holdols.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
