<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="genstats">
<title>Simulating data â€¢ genstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating data">
<meta property="og:description" content="genstats">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">genstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Simulation.html">Simulating data</a>
    <a class="dropdown-item" href="../articles/GWAS.html">GWAS</a>
    <a class="dropdown-item" href="../articles/Gibbs-sampler.html">Gibbs sampler</a>
    <a class="dropdown-item" href="../articles/LT-FH.html">LT-FH</a>
    <a class="dropdown-item" href="../articles/Prediction.html">Prediction</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="https://github.com/Holdols/genstats">
    <span class="fa fa-github fa"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulating data</h1>
            
      
      
      <div class="d-none name"><code>Simulation.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://holdols.github.io/genstats" class="external-link">genstats</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigsnpr/" class="external-link">bigsnpr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'bigsnpr' was built under R version 4.0.5</span>
<span class="co">#&gt; Warning: package 'bigstatsr' was built under R version 4.0.5</span></code></pre></div>
<div class="section level2">
<h2 id="simulation-without-family-history">Simulation without family history<a class="anchor" aria-label="anchor" href="#simulation-without-family-history"></a>
</h2>
<p>The package offers different methods of simulating data. Fist we are
going to show how to simply get out the genotypes of N subjects and the
genetic and ful liabilities. We are here using a small scale test, but
we recomend that if the data should be used for actual statistc analysis
that at least <span class="math inline">\(10,000\times10,000\)</span>
datapoints are simulated.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gen_sim.html">gen_sim</a></span><span class="op">(</span><span class="st">'simple_sim_example'</span>, N<span class="op">=</span><span class="fl">1000</span>, M<span class="op">=</span><span class="fl">1000</span>, C<span class="op">=</span><span class="fl">50</span>, block_size <span class="op">=</span> <span class="fl">100</span>, fam<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Now there will exist a rds file in the working directory with the
data, which we will load.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simple_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">'simple_sim_example.rds'</span><span class="op">)</span>
<span class="va">simple_data</span>
<span class="co">#&gt; $genotypes</span>
<span class="co">#&gt; A Filebacked Big Matrix of type 'code 256' with 1000 rows and 1000 columns.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $map</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,000 x 3</span></span>
<span class="co">#&gt;      snp  beta    MAF</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1     0 0.241 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2     0 0.073<span style="text-decoration: underline;">9</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3     0 0.138 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4     0 0.110 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5     0 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6     0 0.219 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7     0 0.217 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8     0 0.318 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9     0 0.387 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10     0 0.113 </span>
<span class="co">#&gt; <span style="color: #949494;"># ... with 990 more rows</span></span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $fam</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,000 x 4</span></span>
<span class="co">#&gt;      FID    l_g_0   l_f_0 pheno_0</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 -<span style="color: #BB0000;">0.360</span>   -<span style="color: #BB0000;">0.069</span><span style="color: #BB0000; text-decoration: underline;">5</span>       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2  1.22     0.827        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3 -<span style="color: #BB0000;">0.007</span><span style="color: #BB0000; text-decoration: underline;">64</span> -<span style="color: #BB0000;">0.366</span>        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4  0.308    1.38         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5  0.087<span style="text-decoration: underline;">2</span>  -<span style="color: #BB0000;">1.28</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6 -<span style="color: #BB0000;">0.071</span><span style="color: #BB0000; text-decoration: underline;">9</span>   0.256        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7 -<span style="color: #BB0000;">1.61</span>    -<span style="color: #BB0000;">2.57</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8 -<span style="color: #BB0000;">1.32</span>    -<span style="color: #BB0000;">1.35</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9 -<span style="color: #BB0000;">1.23</span>    -<span style="color: #BB0000;">1.08</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10 -<span style="color: #BB0000;">2.00</span>    -<span style="color: #BB0000;">2.61</span>         0</span>
<span class="co">#&gt; <span style="color: #949494;"># ... with 990 more rows</span></span></code></pre></div>
<p>Looking at the data we see that the file contains a list. We can here
find the genotypes (G), liabilities (fam) and information for each SNP
(map).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">simple_data</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span>
<span class="co">#&gt; [1,]    1    1    2    0    0</span>
<span class="co">#&gt; [2,]    1    0    0    0    0</span>
<span class="co">#&gt; [3,]    0    0    0    0    1</span>
<span class="co">#&gt; [4,]    1    0    0    0    0</span>
<span class="co">#&gt; [5,]    0    1    0    0    0</span>
<span class="va">simple_data</span><span class="op">$</span><span class="va">fam</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,000 x 4</span></span>
<span class="co">#&gt;      FID    l_g_0   l_f_0 pheno_0</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 -<span style="color: #BB0000;">0.360</span>   -<span style="color: #BB0000;">0.069</span><span style="color: #BB0000; text-decoration: underline;">5</span>       0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2  1.22     0.827        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3 -<span style="color: #BB0000;">0.007</span><span style="color: #BB0000; text-decoration: underline;">64</span> -<span style="color: #BB0000;">0.366</span>        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4  0.308    1.38         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5  0.087<span style="text-decoration: underline;">2</span>  -<span style="color: #BB0000;">1.28</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6 -<span style="color: #BB0000;">0.071</span><span style="color: #BB0000; text-decoration: underline;">9</span>   0.256        0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7 -<span style="color: #BB0000;">1.61</span>    -<span style="color: #BB0000;">2.57</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8 -<span style="color: #BB0000;">1.32</span>    -<span style="color: #BB0000;">1.35</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9 -<span style="color: #BB0000;">1.23</span>    -<span style="color: #BB0000;">1.08</span>         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10 -<span style="color: #BB0000;">2.00</span>    -<span style="color: #BB0000;">2.61</span>         0</span>
<span class="co">#&gt; <span style="color: #949494;"># ... with 990 more rows</span></span>
<span class="va">simple_data</span><span class="op">$</span><span class="va">map</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,000 x 3</span></span>
<span class="co">#&gt;      snp  beta    MAF</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1     0 0.241 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     2     0 0.073<span style="text-decoration: underline;">9</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     3     0 0.138 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4     0 0.110 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     5     0 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     6     0 0.219 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     7     0 0.217 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     8     0 0.318 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     9     0 0.387 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>    10     0 0.113 </span>
<span class="co">#&gt; <span style="color: #949494;"># ... with 990 more rows</span></span></code></pre></div>
<p>Here the effect of a each causal snip is of the distribution <span class="math inline">\(\mathcal{N}\left( 0, \frac{h^2}{c}\right)\)</span>
and the each minor allele is of the distribution <span class="math inline">\(unif(0.01, 0.49)\)</span>. Using the function
dist_check() we can see if the distribution of the full liabilities is
as expected, which is standard normal. We are here using af small test
set and we cannot expect the the mean and variance to be close to 0 and
1.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dist_check.html">dist_check</a></span><span class="op">(</span><span class="va">simple_data</span><span class="op">)</span></code></pre></div>
<p><img src="Simulation_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code><span class="co">#&gt;     mean_l0 variance_l0 </span>
<span class="co">#&gt;   -0.552989    1.634673</span></code></pre>
<p>We can see that the mean and variance differ a lot from 0 and 1, as
stated this a a very small dataset but with a larger simulation the
distribution will come a lot closer to the expected values, the qqplot
however resembles a normal distribution. It is now possible to perfom
GWAS for association analysis or use the genotypes for prediction.</p>
</div>
<div class="section level2">
<h2 id="simulation-with-family-history">Simulation with family history<a class="anchor" aria-label="anchor" href="#simulation-with-family-history"></a>
</h2>
<p>Simulation with family history is very similar to simulating
without.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gen_sim.html">gen_sim</a></span><span class="op">(</span><span class="st">'genetic_data'</span>, N<span class="op">=</span><span class="fl">10000</span>, M<span class="op">=</span><span class="fl">10000</span>, C<span class="op">=</span><span class="fl">100</span>, block_size <span class="op">=</span> <span class="fl">1000</span>, n_sib<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">family_data</span> <span class="op">=</span> <span class="fu"><a href="../reference/snp_attach.html">snp_attach</a></span><span class="op">(</span><span class="st">'genetic_data.rds'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">family_data</span><span class="op">$</span><span class="va">fam</span><span class="op">)</span>
<span class="co">#&gt;   FID      l_f_0      l_g_0 pheno_0      l_f_p1     l_g_p1 pheno_p1      l_f_p2</span>
<span class="co">#&gt; 1   1 -0.3384807 -0.7064320       0  0.86542968 -0.4871935        0  0.10331741</span>
<span class="co">#&gt; 2   2  1.4324619  0.4944826       0  0.77872081  1.1045429        0  0.60545412</span>
<span class="co">#&gt; 3   3 -1.0801217 -0.3403101       0 -0.09001458  1.0320972        0 -0.22832639</span>
<span class="co">#&gt; 4   4  1.2669931  0.5825709       0 -1.24249284 -0.3447710        0  0.86003095</span>
<span class="co">#&gt; 5   5  0.3125711  1.0115020       0  1.74724276  1.3767706        1  0.87617390</span>
<span class="co">#&gt; 6   6 -0.5124047  0.1266217       0 -0.42144219 -0.1451376        0 -0.02902428</span>
<span class="co">#&gt;       l_g_p2 pheno_p2     l_f_s3     l_g_s3 pheno_s3      l_f_s4      l_g_s4</span>
<span class="co">#&gt; 1  0.2801320        0  0.5112359  0.2131810        0 -1.50060122 -0.11247134</span>
<span class="co">#&gt; 2  0.4630299        0  2.0563382  1.3112339        1  1.35933840  1.00005536</span>
<span class="co">#&gt; 3 -0.7587923        0 -0.1088843 -0.1558196        0 -0.72531820 -0.79113094</span>
<span class="co">#&gt; 4  1.2059687        0  1.5115114  0.4552358        0  0.45513811  0.08464925</span>
<span class="co">#&gt; 5  0.3278623        0  0.4358816  0.9360237        0  0.70031543  1.14029948</span>
<span class="co">#&gt; 6 -0.6010279        0 -0.4068985 -0.5806311        0  0.09545961  0.90836585</span>
<span class="co">#&gt;   pheno_s4</span>
<span class="co">#&gt; 1        0</span>
<span class="co">#&gt; 2        0</span>
<span class="co">#&gt; 3        0</span>
<span class="co">#&gt; 4        0</span>
<span class="co">#&gt; 5        0</span>
<span class="co">#&gt; 6        0</span></code></pre></div>
<p>We see here that we have a bit more family history, but the same same
values each family member as we have for the subject. Again can we check
the distrubition of each full liability using the same function as
before.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dist_check.html">dist_check</a></span><span class="op">(</span><span class="va">family_data</span><span class="op">)</span></code></pre></div>
<p><img src="Simulation_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<pre><code><span class="co">#&gt;     mean_l0 variance_l0 </span>
<span class="co">#&gt; 0.007328894 0.944272064 </span>
<span class="co">#&gt;     mean_l1 variance_l1 </span>
<span class="co">#&gt; 0.005552209 1.046033033 </span>
<span class="co">#&gt;     mean_l2 variance_l2 </span>
<span class="co">#&gt;  0.01255945  1.01357212 </span>
<span class="co">#&gt;     mean_l3 variance_l3 </span>
<span class="co">#&gt; -0.02114777  0.96889338 </span>
<span class="co">#&gt;     mean_l4 variance_l4 </span>
<span class="co">#&gt;  -0.0031872   0.9506856</span></code></pre>
<p>Now that a larger dataset have been simulated the mean and variance
is a lot closer to the expected values of the liabilities and the
qqplots for each liability clearly resembles a normal distribution. Now
this data is ready to be analysied using LTFH or used in predictive
analysis.</p>
<div class="section level3">
<h3 id="note-on-parrellelisation-block-size-and-memory-use-">Note on parrellelisation, block size and memory use.<a class="anchor" aria-label="anchor" href="#note-on-parrellelisation-block-size-and-memory-use-"></a>
</h3>
<p>The functions for the simualation have been optimized with
parallelization and do this with all available cores as a standard.
Depending on the computer running the simulation, the size of the
simulation, the block size and amount of RAM this will take up a lot of
memory of the computer. If a user specified amount of cores is
preffered, this can be specified by running options(mc.cores =
core_amount) where core_amount is the amount of cores available to R.
Furthermore depending on the specified block size R will need to
allocate quite a bit of memory, this can create an error on some
computers depending on the RAM, it could be relevant to check <a href="https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/memory.size" class="external-link uri">https://www.rdocumentation.org/packages/utils/versions/3.6.2/topics/memory.size</a>.
If this doesnâ€™t fix the error, decreasing the block size and/or amount
of cores is recommended.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Louise MÃ¸lgaard Laustsen, Ole Sahlholdt Hansen, Holdols.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
